File: src/main/java/com/koddy/server/global/utils/redis/StringRedisOperator.java
package com.koddy.server.global.utils.redis;

import lombok.RequiredArgsConstructor;
import org.springframework.data.redis.core.StringRedisTemplate;
import org.springframework.stereotype.Component;

import java.time.Duration;
import java.util.concurrent.TimeUnit;

@Component
@RequiredArgsConstructor
public class StringRedisOperator implements RedisOperator<String, String> {
    private final StringRedisTemplate redisTemplate;

    @Override
    public void save(final String key, final String value, final long timeout, final TimeUnit timeUnit) {
        redisTemplate.opsForValue().set(key, value, timeout, timeUnit);
    }

    @Override
    public void save(final String key, final String value, final Duration duration) {
        redisTemplate.opsForValue().set(key, value, duration);
    }

    @Override
    public String get(final String key) {
        return redisTemplate.opsForValue().get(key);
    }

    @Override
    public boolean contains(final String key) {
        return redisTemplate.opsForValue().get(key) != null;
    }

    @Override
    public void delete(final String key) {
        redisTemplate.delete(key);
    }

    @Override
    public long incr(final String key) {
        return redisTemplate.opsForValue().increment(key);
    }

    @Override
    public long incr(final String key, final long timeout, final TimeUnit timeUnit) {
        final long count = redisTemplate.opsForValue().increment(key);
        if (count == 1) {
            redisTemplate.opsForValue().getAndExpire(key, timeout, timeUnit);
        }
        return count;
    }

    @Override
    public long incr(final String key, final Duration duration) {
        final long count = redisTemplate.opsForValue().increment(key);
        if (count == 1) {
            redisTemplate.opsForValue().getAndExpire(key, duration);
        }
        return count;
    }
}

File: src/main/java/com/koddy/server/global/utils/redis/StringRedisOperator.kt
package com.koddy.server.global.utils.redis

import com.koddy.server.global.exception.GlobalException
import com.koddy.server.global.exception.GlobalExceptionCode.UNEXPECTED_SERVER_ERROR
import org.springframework.data.redis.core.StringRedisTemplate
import org.springframework.data.redis.core.ValueOperations
import org.springframework.stereotype.Component
import java.time.Duration
import java.util.concurrent.TimeUnit

@Component
class StringRedisOperator(
    private val redisTemplate: StringRedisTemplate,
) : RedisOperator<String, String?> {
    private val executor: ValueOperations<String, String> = redisTemplate.opsForValue()

    override fun save(
        key: String,
        value: String?,
        timeout: Long,
        timeUnit: TimeUnit,
    ) {
        if (value.isNullOrEmpty()) {
            throw GlobalException(UNEXPECTED_SERVER_ERROR)
        }
        executor.set(key, value, timeout, timeUnit)
    }

    override fun save(
        key: String,
        value: String?,
        duration: Duration,
    ) {
        if (value.isNullOrEmpty()) {
            throw GlobalException(UNEXPECTED_SERVER_ERROR)
        }
        executor.set(key, value, duration)
    }

    override fun get(key: String): String? = executor.get(key)

    override fun contains(key: String): Boolean = !executor.get(key).isNullOrBlank()

    override fun delete(key: String): Boolean = redisTemplate.delete(key)

    override fun incr(key: String): Long = executor.increment(key)!!

    override fun incr(
        key: String,
        timeout: Long,
        timeUnit: TimeUnit,
    ): Long {
        val count: Long = executor.increment(key)!!
        if (count == 1L) {
            executor.getAndExpire(key, timeout, timeUnit)
        }
        return count
    }

    override fun incr(
        key: String,
        duration: Duration,
    ): Long {
        val count: Long = executor.increment(key)!!
        if (count == 1L) {
            executor.getAndExpire(key, duration)
        }
        return count
    }
}
